File: go/chapter_3/01_three_stacks.go
package chapter_3

import "fmt"

type MultiStack struct {
	stacks    int
	stackSize int
	stack     []int
	sizes     []int
}

func GetMultiStack(stacks, stackSize int) *MultiStack {
	return &MultiStack{
		stacks,
		stackSize,
		make([]int, stacks*stackSize),
		make([]int, stacks),
	}
}

func (s *MultiStack) Push(stack, value int) error {
	if s.IsFull(stack) {
		return fmt.Errorf("Cannot push. Stack %d is full.", stack)
	}
	s.sizes[stack-1]++
	s.stack[s.Index(stack)] = value
	return nil
}

func (s *MultiStack) Pop(stack int) (int, error) {
	if s.IsEmpty(stack) {
		return -1, fmt.Errorf("Cannot pop. Stack %d is empty.", stack)
	}
	ind := s.Index(stack)
	ret := s.stack[ind]
	s.stack[ind] = 0
	s.sizes[stack-1]--
	return ret, nil
}

func (s MultiStack) Peek(stack int) (int, error) {
	if s.IsEmpty(stack) {
		return -1, fmt.Errorf("Cannot peek. Stack %d is empty.", stack)
	}
	return s.stack[s.Index(stack)], nil
}

func (s MultiStack) Index(stack int) int {
	stack--
	return stack*s.stackSize + s.sizes[stack] - 1
}

func (s MultiStack) IsEmpty(stack int) bool {
	return s.sizes[stack-1] == 0
}

func (s MultiStack) IsFull(stack int) bool {
	return s.sizes[stack-1] == s.stackSize
}

File: python/chapter_3/01_three_stacks.py
# Опишите, как бы вы использовали один одномерный массив для реализации
# трех стеков
class MultiStack:
    def __init__(self, stack_size):
        self.number_of_stacks = 3
        self.stack = [0] * (self.number_of_stacks * stack_size)
        self.sizes = [0] * self.number_of_stacks
        self.stack_size = stack_size

    def __str__(self):
        return (
            f"number_of_stacks={self.number_of_stacks}\n"
            f"stack={self.stack}\n"
            f"sizes={self.sizes}\n"
            f"stack_size={self.stack_size}\n"
        )

    def push(self, stack_num, value):
        if self.is_full(stack_num):
            raise Exception("Stack is full")
        self.sizes[stack_num] += 1
        self.stack[self.index(stack_num)] = value

    def pop(self, stack_num):
        if self.is_empty(stack_num):
            raise Exception("Stack is empty")
        ind = self.index(stack_num)
        ret, self.stack[ind] = self.stack[ind], 0
        self.sizes[stack_num] -= 1
        return ret

    def peek(self, stack_num):
        if self.is_empty(stack_num):
            raise Exception("Stack is empty")
        return self.stack[self.index(stack_num)]

    def is_empty(self, stack_num):
        return self.sizes[stack_num] == 0

    def is_full(self, stack_num):
        return self.sizes[stack_num] == self.stack_size

    def index(self, stack_num):
        return stack_num * self.stack_size + self.sizes[stack_num] - 1


if __name__ == "__main__":
    s = MultiStack(2)
    s.push(0, 1)
    s.push(0, 2)
    print(s)
    try:
        s.push(0, 3)
    except Exception as e:
        print(e)
    print(s.is_empty(1))
    s.push(1, 10)
    s.push(1, 20)
    s.push(2, 100)
    s.push(2, 200)
    print(s)
    print(s.peek(2))
    print(s.pop(1))
    print(s.pop(1))
    try:
        s.pop(1)
    except Exception as e:
        print(e)
    print(s.is_empty(1))
    print(s)
